src/lsp/symbols.rs

```diff
@@ -4,45 +4,84 @@ use crate::{
     quickfix_list::Location,
 };
 use lsp_types::{DocumentSymbolResponse, SymbolKind};
-use shared::icons::get_icon_config;
+use shared::{canonicalized_path::CanonicalizedPath, icons::get_icon_config};
 
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub(crate) struct Symbols {
     pub(crate) symbols: Vec<Symbol>,
 }
 
-impl TryFrom<DocumentSymbolResponse> for Symbols {
-    type Error = anyhow::Error;
+fn collect_document_symbols(
+    document_symbol: &lsp_types::DocumentSymbol,
+    symbols: &mut Vec<Symbol>,
+    parent_name: Option<String>,
+    path: &CanonicalizedPath,
+) -> Result<(), anyhow::Error> {
+    let mut symbol = Symbol::try_from_document_symbol(document_symbol.clone(), path.clone())?;
+    symbol.container_name = parent_name.clone(); // Set the container_name
+    symbols.push(symbol);
 
-    fn try_from(value: DocumentSymbolResponse) -> Result<Self, Self::Error> {
+    if let Some(children) = document_symbol.clone().children {
+        for child in children {
+            collect_document_symbols(&child, symbols, Some(document_symbol.name.clone()), path)?;
+        }
+    };
+
+    Ok(())
+}
+
+impl Symbols {
+    pub(crate) fn try_from_document_symbol_response(
```

Comment:
Making `collect_document_symbols` pure would reduce the necessary changes in this function.